File: legacy/src/scenes/dungeon/armor-styles.ts
/**
 * ARMOR STYLES — elemental plate SETS sold at the Tavern Armorer.
 *
 * A style is a permanent unlock (localStorage, legacy.ts pattern — it survives
 * death like wallet gold does) that re-skins the knight's plate everywhere the
 * sprite is painted: dungeon, walkable tavern, multiball echoes, paperdolls.
 * One style is ACTIVE at a time; "iron" is the free classic default.
 *
 * Tuning intent (matches the legacy-perk philosophy): these are PRESTIGE
 * purchases. At GOLD_PER_KILL=2 a set costs hundreds of kills — several runs of
 * banked savings, not an impulse buy. The mechanical edge is real but modest:
 * while an elemental style is worn, the Armorer's plate is finer steel
 * (helmet/armor soak a little more per purchase). The LOOK is the product.
 *
 * Pure data + fail-soft persistence — DOM/three-free so it unit-tests headlessly.
 */

export type ArmorStyleId = "iron" | "ice" | "wind" | "fire" | "thunder";

export interface ArmorStyleDef {
  id: ArmorStyleId;
  label: string;
  icon: string;
  /** Wallet gold to unlock, forever. 0 = always owned (iron). */
  price: number;
  /** UI swatch (armorer row chip) — matches the sprite's plate mid tone. */
  swatch: string;
  blurb: string;
  /** Extra soak the Armorer's plate carries while this style is worn. */
  bonusAbsorb: { helmet: number; armor: number };
}

export const ARMOR_STYLES: Record<ArmorStyleId, ArmorStyleDef> = {
  iron: { id: "iron", label: "Crypt Iron", icon: "🛡️", price: 0, swatch: "#8a94a6", blurb: "the classic plate you marched in with", bonusAbsorb: { helmet: 0, armor: 0 } },
  ice: { id: "ice", label: "Glacier Plate", icon: "❄️", price: 600, swatch: "#6fd0e8", blurb: "hoarfrost steel, cold-blue sheen", bonusAbsorb: { helmet: 2, armor: 3 } },
  wind: { id: "wind", label: "Gale Plate", icon: "🌪️", price: 600, swatch: "#8fc46b", blurb: "jade-green tempest steel", bonusAbsorb: { helmet: 2, armor: 3 } },
  fire: { id: "fire", label: "Ember Plate", icon: "🔥", price: 750, swatch: "#f0a63c", blurb: "forge-hot plate, ember glow", bonusAbsorb: { helmet: 2, armor: 3 } },
  thunder: { id: "thunder", label: "Storm Plate", icon: "⚡", price: 900, swatch: "#ffd98a", blurb: "storm-slate chased with lightning gold", bonusAbsorb: { helmet: 2, armor: 3 } },
};

export const ARMOR_STYLE_IDS: ArmorStyleId[] = ["iron", "ice", "wind", "fire", "thunder"];
/** The purchasable elemental sets, in shop order. */
export const ELEMENTAL_STYLE_IDS: ArmorStyleId[] = ["ice", "wind", "fire", "thunder"];

function isStyleId(v: unknown): v is ArmorStyleId {
  return typeof v === "string" && (ARMOR_STYLE_IDS as string[]).includes(v);
}

// ── Persistence (best-depth.ts pattern: one key, shape-validated, fail-soft) ──

const KEY = "pinball-knight-armor-styles";

interface StyleState {
  unlocked: ArmorStyleId[];
  active: ArmorStyleId;
}

let cached: StyleState | null = null;

function load(): StyleState {
  if (cached) return cached;
  const out: StyleState = { unlocked: [], active: "iron" };
  try {
    const raw = localStorage.getItem(KEY);
    if (raw) {
      const p = JSON.parse(raw) as Record<string, unknown>;
      if (Array.isArray(p.unlocked)) out.unlocked = p.unlocked.filter(isStyleId);
      if (isStyleId(p.active)) out.active = p.active;
    }
  } catch (_e) {
    // Blocked storage → session-only styles; buying still works in-memory.
  }
  // Active must be something you own — a corrupt blob never paints a locked set.
  if (out.active !== "iron" && !out.unlocked.includes(out.active)) out.active = "iron";
  cached = out;
  return out;
}

function save(): void {
  try {
    localStorage.setItem(KEY, JSON.stringify(cached ?? { unlocked: [], active: "iron" }));
  } catch (_e) {
    // Session-only is fine.
  }
}

/** The style the knight paints in right now ("iron" until a set is worn). */
export function activeStyle(): ArmorStyleId {
  return load().active;
}

export function isStyleUnlocked(id: ArmorStyleId): boolean {
  return id === "iron" || load().unlocked.includes(id);
}

/** Record a bought set (caller pays the gold) and wear it immediately. */
export function unlockStyle(id: ArmorStyleId): void {
  const st = load();
  if (id !== "iron" && !st.unlocked.includes(id)) st.unlocked.push(id);
  st.active = id;
  save();
}

/** Wear an owned style (no-op on a locked one). Returns whether it applied. */
export function setActiveStyle(id: ArmorStyleId): boolean {
  if (!isStyleUnlocked(id)) return false;
  const st = load();
  st.active = id;
  save();
  return true;
}

/**
 * Soak the Armorer grants when selling `slot` plate under the worn style —
 * base GEAR soak plus the style's finer-steel bonus (boots keep the `1`
 * equipped-sentinel; they never absorb). Floor-found plate stays base iron
 * values regardless of style — the style is a skin over whatever you wear.
 */
export function styleGearGrant(slot: "helmet" | "armor" | "boots", base: number, id: ArmorStyleId = activeStyle()): number {
  if (slot === "boots") return base;
  return base + ARMOR_STYLES[id].bonusAbsorb[slot];
}

/** Test seam. */
export function __resetArmorStylesCache(): void {
  cached = null;
}

File: Rust/crates/pk-core/src/armor_styles.rs
//! ARMOR STYLES — elemental plate SETS sold at the Tavern Armorer.
//!
//! Port of `legacy/src/game/pinball-knight/armor-styles.ts` (128 lines).
//!
//! PORTS: `armor-styles.ts`

use std::sync::RwLock;

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum ArmorStyleId {
    Iron,
    Ice,
    Wind,
    Fire,
    Thunder,
}

impl ArmorStyleId {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Iron => "iron",
            Self::Ice => "ice",
            Self::Wind => "wind",
            Self::Fire => "fire",
            Self::Thunder => "thunder",
        }
    }

    pub fn from_str_id(s: &str) -> Option<Self> {
        match s {
            "iron" => Some(Self::Iron),
            "ice" => Some(Self::Ice),
            "wind" => Some(Self::Wind),
            "fire" => Some(Self::Fire),
            "thunder" => Some(Self::Thunder),
            _ => None,
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct BonusAbsorb {
    pub helmet: i32,
    pub armor: i32,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct ArmorStyleDef {
    pub id: ArmorStyleId,
    pub label: &'static str,
    pub icon: &'static str,
    pub price: i64,
    pub swatch: &'static str,
    pub blurb: &'static str,
    pub bonus_absorb: BonusAbsorb,
}

pub const ARMOR_STYLE_IDS: [ArmorStyleId; 5] = [
    ArmorStyleId::Iron,
    ArmorStyleId::Ice,
    ArmorStyleId::Wind,
    ArmorStyleId::Fire,
    ArmorStyleId::Thunder,
];

pub const ELEMENTAL_STYLE_IDS: [ArmorStyleId; 4] = [
    ArmorStyleId::Ice,
    ArmorStyleId::Wind,
    ArmorStyleId::Fire,
    ArmorStyleId::Thunder,
];

pub const ARMOR_STYLES: [(ArmorStyleId, ArmorStyleDef); 5] = [
    (
        ArmorStyleId::Iron,
        ArmorStyleDef {
            id: ArmorStyleId::Iron,
            label: "Crypt Iron",
            icon: "🛡️",
            price: 0,
            swatch: "#8a94a6",
            blurb: "the classic plate you marched in with",
            bonus_absorb: BonusAbsorb {
                helmet: 0,
                armor: 0,
            },
        },
    ),
    (
        ArmorStyleId::Ice,
        ArmorStyleDef {
            id: ArmorStyleId::Ice,
            label: "Glacier Plate",
            icon: "❄️",
            price: 600,
            swatch: "#6fd0e8",
            blurb: "hoarfrost steel, cold-blue sheen",
            bonus_absorb: BonusAbsorb {
                helmet: 2,
                armor: 3,
            },
        },
    ),
    (
        ArmorStyleId::Wind,
        ArmorStyleDef {
            id: ArmorStyleId::Wind,
            label: "Gale Plate",
            icon: "🌪️",
            price: 600,
            swatch: "#8fc46b",
            blurb: "jade-green tempest steel",
            bonus_absorb: BonusAbsorb {
                helmet: 2,
                armor: 3,
            },
        },
    ),
    (
        ArmorStyleId::Fire,
        ArmorStyleDef {
            id: ArmorStyleId::Fire,
            label: "Ember Plate",
            icon: "🔥",
            price: 750,
            swatch: "#f0a63c",
            blurb: "forge-hot plate, ember glow",
            bonus_absorb: BonusAbsorb {
                helmet: 2,
                armor: 3,
            },
        },
    ),
    (
        ArmorStyleId::Thunder,
        ArmorStyleDef {
            id: ArmorStyleId::Thunder,
            label: "Storm Plate",
            icon: "⚡",
            price: 900,
            swatch: "#ffd98a",
            blurb: "storm-slate chased with lightning gold",
            bonus_absorb: BonusAbsorb {
                helmet: 2,
                armor: 3,
            },
        },
    ),
];

pub fn armor_style_def(id: ArmorStyleId) -> ArmorStyleDef {
    for (style_id, def) in ARMOR_STYLES {
        if style_id == id {
            return def;
        }
    }
    ARMOR_STYLES[0].1
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ArmorStyleState {
    pub unlocked: Vec<ArmorStyleId>,
    pub active: ArmorStyleId,
}

impl Default for ArmorStyleState {
    fn default() -> Self {
        Self {
            unlocked: vec![ArmorStyleId::Iron],
            active: ArmorStyleId::Iron,
        }
    }
}

static STYLE_STATE: RwLock<Option<ArmorStyleState>> = RwLock::new(None);

pub fn active_style() -> ArmorStyleId {
    let read = STYLE_STATE.read().unwrap();
    if let Some(state) = read.as_ref() {
        state.active
    } else {
        ArmorStyleId::Iron
    }
}

pub fn is_style_unlocked(id: ArmorStyleId) -> bool {
    if id == ArmorStyleId::Iron {
        return true;
    }
    let read = STYLE_STATE.read().unwrap();
    if let Some(state) = read.as_ref() {
        state.unlocked.contains(&id)
    } else {
        false
    }
}

pub fn unlock_style(id: ArmorStyleId) {
    let mut write = STYLE_STATE.write().unwrap();
    let state = write.get_or_insert_with(ArmorStyleState::default);
    if id != ArmorStyleId::Iron && !state.unlocked.contains(&id) {
        state.unlocked.push(id);
    }
    state.active = id;
}

pub fn set_active_style(id: ArmorStyleId) -> bool {
    if !is_style_unlocked(id) {
        return false;
    }
    let mut write = STYLE_STATE.write().unwrap();
    let state = write.get_or_insert_with(ArmorStyleState::default);
    state.active = id;
    true
}

pub fn style_gear_grant(slot: &str, base: i32, id: ArmorStyleId) -> i32 {
    let def = armor_style_def(id);
    match slot {
        "helmet" => base + def.bonus_absorb.helmet,
        "armor" => base + def.bonus_absorb.armor,
        _ => base,
    }
}

pub fn reset_armor_styles_cache() {
    let mut write = STYLE_STATE.write().unwrap();
    *write = None;
}
